# DEBUG

- 错误处理

  - `try ... except ... finally`
    - 错误属于类，所有错误类型继承自`BaseException`，所以在使用时注意，它不但捕获该类型的错误，还把其子类也一并捕获。
    - 不必在每个可能出错的地方捕获错误，可以在更高层级捕获（如，在出错的函数外部捕获）。

  ```python
  try:
      print('try...')
      r = 10 / 0
      print('reault = ', r)
  except ZeroDivisionError as e:
      print('except:', e)
  else:
      print('no error!')
  finally:
      print('finally ...')
  print('END')
  
  # try ...
  # except: division by zero
  # finally ...
  # END
  ```

  - 调用栈
    - 如果错误未被捕获，它会一直往上抛，最后呗解释器捕获，打印错误信息，然后程序退出。
    - 所以出错时，通过控制台的调用栈记录可以定位错误。
  - 记录错误
    - 在打印栈记录之后不退出程序。

  ```python
  import logging
  
  # ... same contents ...
  except Exception as e:
      logging.exception(e)
  ```

  - 抛出错误
    - 可以定义自己的一个错误类。
    - 使用`raise`语句进行抛出错误。
    - `raise`语句如果不接具体类型，将继续抛出当前的错误（也就是当前的`exception`的错误类型）。

- 调试

  - print大法

    - 将来程序还得删掉，所以存在许多垃圾信息。

  - 断言

    - `assert xxx, yyy`表示如果xxx为真，否则执行yyy。
    - 如果断言失败，会抛出`AssertionError`。
    - 使用`-o`参数可以关闭断言，如`python3 -o test.py`。

  - logging

    - 不会抛出错误，可以输出栈信息。
    - 可以记录信息的级别，有`debug,info,warning,error`，可以自定设定级别（比设定级别低的消息将不会输出）

    ```python
    import logging
    logging.basicConfig(level=logging.INFO)	# debug messages aren't shown
    ```

    

  - pdb

    - python自带调试器。使用`python3 -m pdb test.py`打开。
    - 输入命令`l`来查看代码。
    - 输入命令`n`单步执行。
    - 输入命令`p `+变量名，可以查看变量的值。
    - 输入命令`q`停止调试。

  - pdb.set_trace()

    - 同样用到pdb，但是不需要单步执行。

    ```python
    import pdb
    
    # codes here...
    a = b + c
    pdb.set_trace()	# program stops here.
    ```

    - 运行到断点处自动进入pdb环境，可以选择`p`查看变量或者`c`继续运行。

  - IDE

    - 如pycharm、vs。

- 测试

  - 单元测试

    - 编写
      - 对一个类、函数、模块的正确性检验。
      - 将测试用例放到一个测试模块中，就是一个完整的单元测试。
      - 编写单元测试时，引入python自带的`unittest`模块。
      - `assertTrue, assertEqual, assertRaise`是三个内置的断言。
      - 以`test`开头的方法是测试方法，否则不会被执行。
    - 运行
      - 最简单的是和以往一样加入`if __name__ == '__main__'`的判断，如下。
      - 另一种方法是`python3 -m unittest xxx_test`，设置参数进行测试。
    - 特殊方法
      - `setUp, tearDown`是两个特殊方法，分别会在每调用一个测试方法的前后指向。
      - 比如，测试数据库的时候，可以在前一个方法中连接数据库，在后一个方法中断开。

    ```python
    import unittest
    
    # test class
    class TestXXX(unittest.TestCase):
        def test_init(self):
            self.assertTrue(...)
            self.assertEqual(f(a), b)
            
        def test_xxx(self):	# test some cases.
            # ...
            
        def test_xxxerror(self):	# test some invalid cases and see if an error occurs.
            # ...
            with self.assertRaise(xxxError):
                # ...
    
                
    if __name__ == '__main__':
        unittest.main()
    ```

  - 文档测试

    - 在类的头部加入注释就可以实现的神奇操作。
    - 模块被执行时，如果注释的测试样例和结果没有错，就不会有输出，否则将会报错。
    - 只有在命令行运行时才会执行`doctest`（`python3 mydict2.py`），如果是被导入则不进行测试。
    - `doctest`非常有用，不但可以用于测试，还能直接作为示例代码。
    - 通过某些文档生成工具，就可以直接将注释部分提取出来，用户看文档时也看到了`doctest`。

    ```python
    # mydict2.py
    class Dict(dict):
        '''
        Simple dict but also support access as x.y style.
    
        >>> d1 = Dict()
        >>> d1['x'] = 100
        >>> d1.x
        100
        >>> d1.y = 200
        >>> d1['y']
        200
        >>> d2 = Dict(a=1, b=2, c='3')
        >>> d2.c
        '3'
        >>> d2['empty']
        Traceback (most recent call last):
            ...
        KeyError: 'empty'
        >>> d2.empty
        Traceback (most recent call last):
            ...
        AttributeError: 'Dict' object has no attribute 'empty'
        '''
        def __init__(self, **kw):
            super(Dict, self).__init__(**kw)
    
        def __getattr__(self, key):
            try:
                return self[key]
            except KeyError:
                raise AttributeError(r"'Dict' object has no attribute '%s'" % key)
    
        def __setattr__(self, key, value):
            self[key] = value
    
    if __name__=='__main__':
        import doctest
        doctest.testmod()
    ```

    